http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java

```diff
@@ -228,6 +244,90 @@ public HttpExecuteResponse call() throws IOException {
                                       .build();
         }
 
+        private Optional<OutputStream> tryGetOutputStream() {
+            return getAndHandle100Bug(() -> invokeSafely(connection::getOutputStream), false);
+        }
+
+        private Optional<InputStream> tryGetInputStream() {
+            return getAndHandle100Bug(() -> invokeSafely(connection::getInputStream), true);
+        }
+
+        private Optional<InputStream> tryGetErrorStream() {
+            InputStream result = invokeSafely(connection::getErrorStream);
+            if (result == null && expect100BugEncountered) {
+                log.debug(() -> "The response payload has been dropped because of a limitation of the JDK's URL Connection "
+                                + "HTTP client, resulting in a less descriptive SDK exception error message. Using "
+                                + "the Apache HTTP client removes this limitation.");
+            }
+            return Optional.ofNullable(result);
+        }
+
+        /**
+         * This handles a bug in {@link HttpURLConnection#getOutputStream()} and {@link HttpURLConnection#getInputStream()}
+         * where these methods will throw a ProtocolException if we sent an "Expect: 100-continue" header, and the
+         * service responds with something other than a 100.
+         *
+         * HttpUrlConnection still gives us access to the response code and headers when this bug is encountered, so our
+         * handling of the bug is:
+         * <ol>
+         *     <li>If the service returned a response status or content length that indicates there was no response payload,
+         *     we ignore that we couldn't read the response payload, and just return the response with what we have.</li>
+         *     <li>If the service returned a payload and we can't read it because of the bug, we throw an exception for
+         *     non-failure cases (2xx, 3xx) or log and return the response without the payload for failure cases (4xx or 5xx)
+         *     .</li>
+         * </ol>
+         */
+        private <T> Optional<T> getAndHandle100Bug(Supplier<T> supplier, boolean failOn100Bug) {
+            try {
+                return Optional.ofNullable(supplier.get());
+            } catch (RuntimeException e) {
+                if (!exceptionCausedBy100HandlingBug(e)) {
+                    throw e;
+                }
+
+                expect100BugEncountered = true;
```

Comment:
Should we put `responseHasNoContent()` before this, so that we only emit warnings when we're actually missing some potential behavior?

Reply:
Sure. responseHasNoContent() is relatively slow to check, so I'll have to rearrange some things to ensure we don't check it twice in the success case.